Emote Moderation System For Servers

Carter4242 opened the pull request:
Allows you to turn a server-side version of Emotecraft into "whitelist mode", which makes it so only emotes that the server admin uploads to the server-side whitelist directory can be used. Emotes that aren't whitelisted will be canceled and won't be displayed to other users.

This is accomplished efficiently by checking the hash of the emote, which is generated by checking all of the emote's unique properties, namely all boneAnimations frames, any sounds, and any particles.

How these properties are structured could easily change from version to version of Emotecraft, so it's only guaranteed that this will work if the client and server are on the same version of Emotecraft. However, from my testing, all 3.0 versions including the latest self-compiled version from build 106 onward have worked fine with this.

Server-side hashes are only taken once, unless an emote is modified, or the emotecraft jar is updated. Hashes are refreshed when the server is started, or via /emotes whitelist reload. You can also force a new hash to be taken of every emote via /emotes whitelist force-reload.

This module does nothing unless you set the new config option enableServerEmoteWhitelist to true. It doesn't do anything for clients.

Carter4242 committed:
intial hash based approach

Carter4242 committed:
Merge pull request #1 from KosmX/dev

merge depend update

Carter4242 committed:
force loadEmotesServerSide to be false, as we don't do anything with it right now

Carter4242 committed:
move the default emotes folder server side to config/emotecraft/emotes

Carter4242 committed:
Remove debug from AbstractServerEmotePlay, and make it check the config instead of the moderation instance

Carter4242 committed:
Add EmoteWhitelistHashManager and Emote moderation functionality

Allows you to turn a server side version of Emotecraft into "whitelist mode", this makes it so only emotes that the server admin uploads to the server-side whitelist directory can be used by players on the server. Emotes that aren't whitelisted will be canceled and won't be displayed to other users.

This is accomplished efficiently by checking the hash of the emote, which is generated by checking all of the emotes unique properties, namely all boneAnimations frames, any sounds, any particles, and the emote loop type.

How these properties are structured could easily change version to version of Emotecraft, so it's only guaranteed that this will work if the client and server are on the same version of Emotecraft.

Server side hashes are only taken once, unless an emote is modified, hashes are refreshed when the server is started, or via /emotes whitelist reload. You can also force a new hash to be taken of every emote via /emotes whitelist force-reload.

Carter4242 committed:
change how code is structued so there are less changes, also we don't need to intalize in fabric/forge anymore, that's handled by mainloader

Carter4242 committed:
change back how we handle EventResults in handleStreamEmote, realistically we should only stop on fail, otherwise it's confusing

Carter4242 committed:
change emotes dir back (it effects clients as well), make whitelist config options clearer (for servers only).

Carter4242 committed:
Log hash when adding an emote for the first time, stop hashing LoopType (breaks things)

Carter4242 committed:
revert loadEmotesServerSide change, as loadEmotes runs on the server as well, so that breaks things

Carter4242 committed:
make whitelistedEmotesDir config title a bit clearer

Carter4242 commented:
Has been tested on both fabric and paper, currently running in production on a paper server. Should work out of the box with Forge as well (I think the startup pipeline for Emotecraft on Forge servers is the same as on Fabric, and almost all logic other than initial load is Emotecraft/Java, not MC).

dima-dencep commented on emotesAPI/src/main/java/io/github/kosmx/emotes/common/SerializableConfig.java:
```diff
@@ -31,6 +31,10 @@ public class SerializableConfig {
 
+    // Emote whitelist configuration
+    public final ConfigEntry<Boolean> enableEmoteWhitelist = new ConfigEntry<>("enableServerEmoteWhitelist", false, false, expert, true);
+    public final ConfigEntry<String> whitelistedEmotesDir = new ConfigEntry<>("emoteServerWhitelistDirectory", "config/emotecraft/whitelist", false, expert, true);
```
Don't store paths like that, there's an `InstanceService` for that

Carter4242 replied:
I think I've fixed this? Switched to using getGameDirectory

dima-dencep replied:
no

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +107,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                        .then(literal("toggle").requires(ctx -> IPermissionService.INSTANCE.check(ctx, "emotes.whitelist.toggle", 4))
+                                .executes(context -> {
+                                    try {
+                                        var config = io.github.kosmx.emotes.server.config.Serializer.getConfig();
```
Fix the import and don't use var

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +107,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                        )
+                        .then(literal("reload").requires(ctx -> IPermissionService.INSTANCE.check(ctx, "emotes.whitelist.reload", 4))
+                                .executes(context -> {
+                                    var config = io.github.kosmx.emotes.server.config.Serializer.getConfig();
```
Fix the import and don't use var

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +107,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                                .executes(context -> {
+                                    var config = io.github.kosmx.emotes.server.config.Serializer.getConfig();
+                                    if (config.enableEmoteWhitelist.get()) {
+                                        io.github.kosmx.emotes.server.moderation.EmoteWhitelistHashManager.setupWhitelistConfig();
```
Fix the import

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +107,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                                        config.enableEmoteWhitelist.set(!currentValue);
+                                        io.github.kosmx.emotes.server.config.Serializer.INSTANCE.saveConfig();
+                                        if (!currentValue) {
+                                            io.github.kosmx.emotes.server.moderation.EmoteWhitelistHashManager.setupWhitelistConfig();
```
Fix the import

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +107,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                        )
+                        .then(literal("force-reload").requires(ctx -> IPermissionService.INSTANCE.check(ctx, "emotes.whitelist.reload", 4))
+                                .executes(context -> {
+                                    var config = io.github.kosmx.emotes.server.config.Serializer.getConfig();
```
Fix the import and don't use var

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteModerator.java:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * Initialize the emote moderator by registering event handlers
+     */
+    public static void initialize() {
+        if (initialized) {
+            return;
+        }
+        
+        // Register the emote verification handler
+        ServerEmoteEvents.EMOTE_VERIFICATION.register(EmoteModerator::verifyEmote);
+        
+        initialized = true;
+    }
```
Move the initialization to clinit and leave the method no-op

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,289 @@
+                }
+
+                Path jarFile = Path.of(jarPath);
+                long jarLastModified = Files.getLastModifiedTime(jarFile).toMillis();
```
Never trust a file's metadata, it is often purged

Carter4242 replied:
Switched to use the emotecraft jar hash. This is not a required feature, it just means it will never do a rehash on update if it fails, I've also updated the warning on failure to reflect that.

dima-dencep commented on minecraft/archCommon/src/main/java/io/github/kosmx/emotes/main/MainLoader.java:
```diff
@@ -21,6 +21,7 @@ public static void main(boolean isClient) {
         } else {
             Serializer.INSTANCE = new Serializer<>(new ConfigSerializer<>(SerializableConfig::new), SerializableConfig.class);
             UniversalEmoteSerializer.loadEmotes();
+            io.github.kosmx.emotes.server.moderation.EmoteWhitelistHashManager.setupWhitelistConfig();
```
Fix the import

dima-dencep commented on paper/src/main/java/io/github/kosmx/emotes/bukkit/BukkitWrapper.java:
```diff
@@ -40,6 +40,7 @@ public void onLoad() {
 
         Serializer.INSTANCE = new Serializer<>(new ConfigSerializer<>(SerializableConfig::new), SerializableConfig.class); //it does register itself
         UniversalEmoteSerializer.loadEmotes();
+        io.github.kosmx.emotes.server.moderation.EmoteWhitelistHashManager.setupWhitelistConfig();
```
Fix the import

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,289 @@
+package io.github.kosmx.emotes.server.moderation;
```
This class should be rewritten to get rid of crappy and useless code

ZigyTheBird commented:
That's a lot of bad imports 🥀XD

Carter4242 committed:
Fix whitelist modules import and var usage in ConfigSeralizer

Carter4242 committed:
Revert "Fix whitelist modules import and var usage in ConfigSeralizer"

This reverts commit 5d34ea2e0267b39864a4d0b252e18c233604d114.

Carter4242 committed:
Fix whitelist modules import and var usage in ServerCommands

Carter4242 committed:
Fixes

EmoteModerator initialize, switch to use file hash for emotecraft version verification, fix dir opening to use InstanceServer, fix imports, delete unused function

Carter4242 committed:
Make EmoteModerator.initialize no op and fix initialize

Carter4242 committed:
fix loader imports to be normal

Carter4242 committed:
Don't log internal copied emotes individually, log all at once at the end

Carter4242 committed:
Replaced array hack with a final local Map for previousHashes

dima-dencep commented on emotesMc/src/main/java/io/github/kosmx/emotes/mc/ServerCommands.java:
```diff
@@ -104,6 +110,62 @@ public static <T> void register(CommandDispatcher<T> dispatcher, boolean isDedic
+                .then(literal("whitelist").requires(ctx -> isDedicated)
+                        .then(literal("toggle").requires(ctx -> IPermissionService.INSTANCE.check(ctx, "emotes.whitelist.toggle", 4))
+                                .executes(context -> {
+                                    try {
+                                        SerializableConfig config = Serializer.getConfig();
+                                        boolean currentValue = config.enableEmoteWhitelist.get();
+                                        config.enableEmoteWhitelist.set(!currentValue);
+                                        Serializer.INSTANCE.saveConfig();
+                                        if (!currentValue) {
+                                            EmoteWhitelistHashManager.setupWhitelistConfig();
+                                        }
+                                        context.getSource().sendSuccess(() -> Component.literal("Emote whitelist " + (!currentValue ? "enabled" : "disabled")), true);
+                                        return 0;
+                                    } catch (Exception e) {
+                                        context.getSource().sendFailure(Component.literal("Failed to toggle whitelist: " + e.getMessage()));
+                                        return -1;
+                                    }
```
Couldn't a boolean argument be used instead of toggle?

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteModerator.java:
```diff
@@ -0,0 +1,44 @@
+package io.github.kosmx.emotes.server.moderation;
```
You've made it so that this class is never initialized, if that's the intention, it's better to remove it

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/network/AbstractServerEmotePlay.java:
```diff
@@ -24,7 +24,7 @@
 public abstract class AbstractServerEmotePlay<P extends IServerNetworkInstance> extends ServerEmoteAPI {
     protected boolean doValidate() {
-        return Serializer.getConfig().validateEmote.get();
+        return (Serializer.getConfig().enableEmoteWhitelist.get() || Serializer.getConfig().validateEmote.get());
```
enableEmoteWhitelist is checked twice

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/network/AbstractServerEmotePlay.java:
```diff
@@ -65,10 +65,10 @@ public void receiveMessage(NetData data, P instance) throws IOException {
     @SuppressWarnings("ConstantConditions")
-    protected void handleStreamEmote(NetData data, P instance) throws IOException {
-        if (!data.valid && doValidate()) {
+    protected void handleStreamEmote(NetData data, P instance) throws IOException {        
```
Remove the spaces at the end of the method

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,291 @@
+    /**
+     * Force refresh: delete emoteHashes.json and rebuild from whitelist folder
+     */
+    public void forceRefreshHashes() {
+        CommonData.LOGGER.warn("Refreshing all emotes for the whitelist [first run or updated Emotecraft], this may take a while if you have many whitelisted emotes.");
+        try {
+            if (Files.exists(hashesFile)) {
+                Files.delete(hashesFile);
+            }
+        } catch (Exception e) {
+            CommonData.LOGGER.warn("Failed to delete emoteHashes.json for force refresh", e);
+        }
+    }
```
Do not save the hashes to a file, just load them with the server load
You can make a separate executor for it, and block the server thread after loading if the hashes have not had time to load

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,291 @@
+
+        fileInfoMap.clear();
+        try {
+            Files.walkFileTree(whitelistDir, new SimpleFileVisitor<Path>() {
```
Remove your emote loading code from here, use emotecraft to do it

Carter4242 committed:
add a getConfigFolder() method to InstanceServerice (and implement for fabric/forge)

Carter4242 committed:
use new getConfigFolder() instead of getGameDirectory

Carter4242 committed:
switch to enable/disable instead of toggle for emotes whitelist

Carter4242 committed:
explicitly register EmoteModerator once when EmoteWhitelistHashManager is first created

Carter4242 committed:
Remove whitelist check from doValidate since it's called in handleStreamEmote already

Carter4242 committed:
formatting

Carter4242 committed:
refactor EmoteWhitelist to not save hashes to file / have persistence

Get rid of jar hashing as well (since we don't care anymore as we always reload)

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteModerator.java:
```diff
@@ -0,0 +1,48 @@
+            ServerEmoteEvents.EMOTE_VERIFICATION.register(EmoteModerator::verifyEmote);
+            registered = true;
+        }
+    }
```
Move the initialization to clinit and leave the method no-op

dima-dencep commented on minecraft/neoforge/src/main/java/io/github/kosmx/emotes/neoforge/executor/ForgeEmotesMain.java:
```diff
@@ -12,9 +12,15 @@ public Path getGameDirectory() {
     public Path getConfigPath() {
-        return FMLPaths.CONFIGDIR.get().resolve("emotecraft.json");
+        return getConfigFolder().resolve("emotecraft.json");
     }
```
If this method is the same as in the superclass, it is better to remove it

dima-dencep commented on minecraft/fabric/src/main/java/io/github/kosmx/emotes/fabric/executor/FabricEmotesMain.java:
```diff
@@ -11,9 +11,15 @@ public Path getGameDirectory() {
     public Path getConfigPath() {
-        return FabricLoader.getInstance().getConfigDir().resolve("emotecraft.json");
+        return getConfigFolder().resolve("emotecraft.json");
     }
```
If this method is the same as in the superclass, it is better to remove it

Carter4242 committed:
Make EmoteModerator register no op and move register to clinit

Carter4242 committed:
remove dupe Fabric/Forge getConfig code

Carter4242 committed:
fix log format

Carter4242 committed:
Fix permission for emotes.whitelist.set (from toggle)

Carter4242 committed:
Merge pull request #2 from KosmX/dev

New Crowdin translations by Github Action (#625)

Carter4242 committed:
Merge remote-tracking branch 'kosmx/dev' into dev

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,203 @@
+    }
+
+
+    public int calculateEmoteHash(Animation emote) {
```
Create junit tests for this

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/services/InstanceService.java:
```diff
@@ -18,23 +18,26 @@ default Path getExternalEmoteDir() {
             }
         }
-        return getGameDirectory().resolve(directoryName).resolve("emotecraft.json");
+        return configDir;
```
```suggestion
        return configDir.resolve(directoryName);
```

Carter4242 replied:
Are you sure that's right? Github diff is weird, that's getConfigFolder(), getConfigPath() is right below it now.

dima-dencep commented on minecraft/fabric/src/main/java/io/github/kosmx/emotes/fabric/executor/FabricEmotesMain.java:
```diff
@@ -11,9 +11,10 @@ public Path getGameDirectory() {
         return FabricLoader.getInstance().getGameDir();
     }
 
+
```
space

dima-dencep commented on minecraft/neoforge/src/main/java/io/github/kosmx/emotes/neoforge/executor/ForgeEmotesMain.java:
```diff
@@ -12,9 +12,10 @@ public Path getGameDirectory() {
         return FMLLoader.getGamePath();
     }
 
+
```
space

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,203 @@
+            allowedHashes.add(info.hash);
+        }
+    }
+
```
space

dima-dencep commented on emotesServer/src/main/java/io/github/kosmx/emotes/server/moderation/EmoteWhitelistHashManager.java:
```diff
@@ -0,0 +1,203 @@
+        } catch (IOException e) {
+            CommonData.LOGGER.warn("Failed to create whitelist emotes directory: {}", whitelistDir.toAbsolutePath(), e);
+        }
+    }
```
Use existing methods for this

Carter4242 replied:
Hey what methods do you mean?

The internal emotes are not written down anywhere, and serializeInternalJson() saves them to memory, which we don't want in this case. I did get the pattern I use here from serializeInternalJson though. I could refactor that file to add a new function they both use if you wanted.

Carter4242 committed:
remove spaces, switch to using UniversalEmoteSerializer.class for the stream instead of EmoteWhitelistHashManager

Carter4242 committed:
start adding tests

Carter4242 committed:
add unit tests for calculateEmoteHash

Carter4242 committed:
Merge branch 'KosmX:dev' into dev

Carter4242 committed:
Merge remote-tracking branch 'upstream/dev' into dev

Carter4242 committed:
Merge remote-tracking branch 'upstream/dev' into dev

Carter4242 committed:
Merge remote-tracking branch 'upstream/dev' into dev

ZigyTheBird committed:
Emote moderation system refactoring
